Automattic/wp-calypso
# Attempt to fix issue where switching flows removes pre-filled plan step data

This PR is to look at a potential fix for https://github.com/Automattic/zelda-private/issues/124 where switching flows removes pre-filled plan step data.

I'm imagining this will likely be a short-term fix as @spen's work in https://github.com/Automattic/wp-calypso/pull/35139 will probably replace this in terms of how the progress state works.

I've also updated the e2e tests in this PR to cover switching to the ecommerce flow and back to the business flow and proceeding to the checkout with the business plan intact. I'm not 100% sure that this PR doesn't introduce any other unintended behaviour, so I'm keen for feedback on the approach.

#### Steps to reproduce the existing issue

Steps to reproduce the behavior on _master_:

1. Go to '/start/business' or '/start/premium'
2. Select 'Online store' at the site type step
3. Navigate back to the site type step
4. Select the `business` site type and continue on the journey to create a site and proceed to checkout
5. The checkout will be empty, and it'll redirect you to the site selection for Plans, but it _should_ be on the checkout with the plan pre-selected.

#### Changes proposed in this Pull Request

* When changing flows, clear out the excluded steps array in `flows` so that those steps will be evaluated again (e.g. `plans-business`). This seems to fix the issue, however then the `plans-business` step gets added in the wrong order in the progress array. To mitigate this issue (and the problems this causes for the `Back` button), this PR also features:
* When skipping the plan step, also set `wasSkipped: true` for that step, so that the `getBackUrl` function method in `NavigationLink` will skip the step. This should fix the back button issue where the `plans-business` step is in the progress state, and it was possible to get stuck in a loop clicking `Back`.
* Update e2e test for the `/start/business` flow to include flow branching (selecting the ecommerce step), clicking "back" and continuing forwards along the original flow, preserving the business plan in the cart.

#### Testing instructions for this fix



1. Go to '/start/business' or '/start/premium'
2. Select 'Online store' at the site type step
3. Enter a domain and proceed to the plans step
3. Navigate all the way back to the site type step
4. Select `business` site type , and proceed all the way to the domains step.
5. Click `back` and get all the way back to the site type step, and you should _not_ get stuck in a loop (the back button should behave correctly)
6. Continue forwards and complete signup with the `business` site type, and you should reach the checkout screen with the correct plan pre-selected
